### PR TITLE
Automatically set `HU.BestBeforeDate` and `HU.LotNr` attributes during Manfacturing

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh/issues/10022

### CODE REVIEW

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/attribute/HUAttributeConstants.java [hunk 1]
```diff
@@ -87,4 +77,7 @@ public static String sqlBestBeforeDate(final String huIdColumnName)
 		Check.assumeNotEmpty(huIdColumnName, "huIdColumnName is not empty");
 		return "\"de.metas.handlingunits\".huBestBeforeDate(" + huIdColumnName + ")";
 	}
+
+	public static final String SYSCONFIG_AutomaticallySetLotNumber = "de.metas.handlingunits.attributes.AutomaticallySetLotNumber";
+	public static final String SYSCONFIG_AutomaticallySetBestBeforeDate = "de.metas.handlingunits.attributes.AutomaticallySetBestBeforeDate";
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/attribute/HUAttributeConstants.java:82]: introduce IHUAttributesBL.isAutomaticallySetLotNumber() and isAutomaticallySetBestBeforeDate().
avoid having public sysconfig names.

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java [hunk 2]
```diff
@@ -334,17 +338,8 @@ private void updateReceivedHUs(
...
-					huAttributes.setValue(AttributeConstants.ATTR_LotNumber, lotNumber);
-				}
-
-				if (bestBeforeDate != null
-						&& huAttributes.hasAttribute(AttributeConstants.ATTR_BestBeforeDate))
-				{
-					huAttributes.setValue(AttributeConstants.ATTR_BestBeforeDate, bestBeforeDate);
-				}
+				setLotNumberAttribute(hu, huAttributes);
+				setBestBeforeDateAttribute(huAttributes);
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java:342]: introduce a new method, e.g. updateAttributes(hu...) which calls both methods.

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java [hunk 3]
```diff
@@ -355,6 +350,33 @@ private void updateReceivedHUs(
 		setAssignedHUs(hus);
 	}
 
+	private void setLotNumberAttribute(final I_M_HU hu, final IAttributeStorage huAttributes)
+	{
+		final String lotNumberToSet;
+		if (Services.get(ISysConfigBL.class).getBooleanValue(HUAttributeConstants.SYSCONFIG_AutomaticallySetLotNumber, false))
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java:356]: use IHUAttributesBL.isAuto....

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java [hunk 4]
```diff
@@ -355,6 +350,33 @@ private void updateReceivedHUs(
...
+		{
+			lotNumberToSet = CoalesceUtil.coalesce(lotNumber, hu.getValue());
+		}
+		else
+		{
+			lotNumberToSet = lotNumber;
+		}
+
+		if (huAttributes.hasAttribute(AttributeConstants.ATTR_LotNumber))
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java:365]: minor: move this as first check

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveCUs.java [hunk 5]
```diff
@@ -206,6 +205,10 @@ private I_M_HU createPlanningVHU(final I_M_ReceiptSchedule receiptSchedule)
 			throw new HUException("One and only one VHU was expected but we got: " + hus);
 		}
 		final I_M_HU vhu = hus.get(0);
+
+		setAttributeBBD(vhu, receiptSchedule);
+		setAttributeLotNumber(vhu);
```
Comment (teosarca) [inline backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveCUs.java:210]: introduce only one method which calls both

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveHUs_Base.java [hunk 6]
```diff
@@ -148,6 +149,11 @@ protected final String doIt() throws Exception
 		//
 		// Generate the HUs
 		final List<I_M_HU> hus = huGenerator.generateWithinOwnTransaction();
+
+		hus.forEach(hu -> {
+			setAttributeBBD(hu, receiptSchedule);
```
Comment (teosarca) [inline backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveHUs_Base.java:154]: introduce one method which calls both

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java [hunk 7]
```diff
@@ -228,7 +232,32 @@ protected void postProcess(boolean success)
 		getViewsRepo().notifyRecordChanged(I_PP_Order.Table_Name, ppOrderLinesView.getPpOrderId().getRepoId());
 	}
 
-	private final IPPOrderReceiptHUProducer newReceiptCandidatesProducer()
+	@Nullable
+	LocalDate computeBestBeforeDate()
+	{
+		if (sysConfigBL.getBooleanValue(HUAttributeConstants.SYSCONFIG_AutomaticallySetBestBeforeDate, false))
```
Comment (teosarca) [inline backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java:238]: use IHUAttributeBL.isAuto...

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java [hunk 8]
```diff
@@ -228,7 +232,32 @@ protected void postProcess(boolean success)
...
+		{
+			final PPOrderLineRow row = getSingleSelectedRow();
+			final I_PP_Order ppOrderPO = huPPOrderBL.getById(row.getOrderId());
+			final LocalDate datePromised = TimeUtil.asLocalDate(ppOrderPO.getDatePromised());
+
+			final ProductId productId = row.getProductId();
+			final int guaranteeDaysMin = productDAO.getProductGuaranteeDaysMinFallbackProductCategory(productId);
+
+			if (guaranteeDaysMin <= 0)
```
Comment (teosarca) [inline backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java:247]: eager make this check

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/attribute/HUAttributeConstants.java (hunk 1 above)
Comment (TheBestPessimist) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/attribute/HUAttributeConstants.java:83]: done

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java (hunk 2 above)
Comment (TheBestPessimist) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java:342]: done

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java (hunk 3 above)
Comment (TheBestPessimist) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java:356]: done

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java (hunk 4 above)
Comment (TheBestPessimist) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/AbstractPPOrderReceiptHUProducer.java:365]: done

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveCUs.java (hunk 5 above)
Comment (TheBestPessimist) [reply backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveCUs.java:210]: done

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveHUs_Base.java (hunk 6 above)
Comment (TheBestPessimist) [reply backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/handlingunits/process/WEBUI_M_ReceiptSchedule_ReceiveHUs_Base.java:154]: done

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java (hunk 7 above)
Comment (TheBestPessimist) [reply backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java:238]: done

File: backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java (hunk 8 above)
Comment (TheBestPessimist) [reply backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pporder/process/WEBUI_PP_Order_Receipt.java:247]: done

**Pull request merged**